model/jpa/src/main/resources/META-INF/jpa-changelog-authz-3.4.0.CR1.xml

```diff
@@ -0,0 +1,93 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<!--
+  ~ * Copyright 2017 Red Hat, Inc. and/or its affiliates
+  ~ * and other contributors as indicated by the @author tags.
+  ~ *
+  ~ * Licensed under the Apache License, Version 2.0 (the "License");
+  ~ * you may not use this file except in compliance with the License.
+  ~ * You may obtain a copy of the License at
+  ~ *
+  ~ * http://www.apache.org/licenses/LICENSE-2.0
+  ~ *
+  ~ * Unless required by applicable law or agreed to in writing, software
+  ~ * distributed under the License is distributed on an "AS IS" BASIS,
+  ~ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ * See the License for the specific language governing permissions and
+  ~ * limitations under the License.
+  -->
+
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.2.xsd">
+    <changeSet author="psilva@redhat.com" id="authz-3.4.0.CR1-resource-server-pk-change">
+        <!-- Data migration to change the PK of RESOURCE_SERVER to use the CLIENT_ID. -->
+        <addColumn tableName="RESOURCE_SERVER_POLICY">
+            <column name="RESOURCE_SERVER_CLIENT_ID" type="VARCHAR(36)"/>
+        </addColumn>
+        <addColumn tableName="RESOURCE_SERVER_RESOURCE">
+            <column name="RESOURCE_SERVER_CLIENT_ID" type="VARCHAR(36)"/>
+        </addColumn>
+        <addColumn tableName="RESOURCE_SERVER_SCOPE">
+            <column name="RESOURCE_SERVER_CLIENT_ID" type="VARCHAR(36)"/>
+        </addColumn>
+
+        <sql>
+            UPDATE RESOURCE_SERVER_POLICY p SET RESOURCE_SERVER_CLIENT_ID=(SELECT CLIENT_ID FROM RESOURCE_SERVER s WHERE s.ID = p.RESOURCE_SERVER_ID);
```

Comment:
This SQL doesn't work on MSSQL:
ERROR [org.keycloak.connections.jpa.updater.liquibase.conn.DefaultLiquibaseConnectionProvider] (ServerService Thread Pool -- 55) Change Set META-INF/jpa-changelog-authz-3.4.0.CR1.xml::authz-3.4.0.CR1-resource-server-pk-change::psilva@redhat.com failed.  Error: Incorrect syntax near 'p'. [Failed SQL: UPDATE RESOURCE_SERVER_POLICY p SET RESOURCE_SERVER_CLIENT_ID=(SELECT CLIENT_ID FROM RESOURCE_SERVER s WHERE s.ID = p.RESOURCE_SERVER_ID)]: liquibase.exception.DatabaseException: Incorrect syntax near 'p'. [Failed SQL: UPDATE RESOURCE_SERVER_POLICY p SET RESOURCE_SERVER_CLIENT_ID=(SELECT CLIENT_ID FROM RESOURCE_SERVER s WHERE s.ID = p.RESOURCE_SERVER_ID)]

Reply:
I will check the syntax, I have a MSSQL instance readily available.

Reply:
You might need to split the current changeset into three, extracting the <sql> part one into a separate one, and making variants of that changeset via <precondition> on <dbms> similarly to what is already used for db2 in https://github.com/hmlnarik/keycloak/blob/master/model/jpa/src/main/resources/META-INF/jpa-changelog-1.8.0.xml#L21-L25

Reply:
This is what I fear. I'm used to create Liquibase changeset twice to support both MSSQL and Oracle.

Reply:
@hmlnarik, if I am to create an exception for a specific DBMS, do you want a separate changelog file for that dbms or two separate changesets in the same file?

Reply:
@glavoie Thanks, two separate changesets in the same file would do.

Reply:
@hmlnarik, will do this then. I've confirmed that 2 changesets will be required. My DBA helped me with tests on multiple DBMS and we can't find a portable solution. For now the exception would be "Oracle", but I would need your help to see what DB2 accepts. Will likely update the PR next Monday.

Reply:
@glavoie Certainly, please just let me know when you consider the PR ready

Reply:
@hmlnarik, just updated the PR again.